### PR TITLE
Performance Improvements

### PR DESCRIPTION
**Pull request by sinlerdev**
I have massively improved CC. 

The first optimization I have immediately made is to create the table with the size of `ThicknessQuality` before working on it:
image
This is because every time you add an item to the table, it reallocates, but with this, it doesn't.

The second optimization I made is to remove the second loop in the `Solver` method, as this increases the time complexity for the whole algorithm and the time that the method overall took *(loops are kinda slow)*. Now the `i` increases by 1 every time the `t` increases if `i` is lower than `self.Quality`  

**EDIT:** I just noticed that the second optimization I made doesn't work as intended. I will search for another mechanism

### CODE REVIEW

**Comment by sinlerdev** [comment]
image
As you can see, my fork is significantly faster with the two optimization I listed before.


code : 
```lua
local SinlerCast = require(game.ReplicatedStorage.SinlerCast)

local CylindricalCast = require(game.ReplicatedStorage.CylindricalCast)

local SinlerClyinder = SinlerCast.new({
	Quality = 30,
	Size = Vector3.new(10, 10, 10),
	Ignore = {},
	ThicknessQuality = 2 
})

local OrginialClyinder = CylindricalCast.new({
	Quality = 30, 
	Size = Vector3.new(10, 10, 10),
	Ignore = {},
	ThicknessQuality = 2
})

local SinlerCast = {}
local OrginialCast = {}


local function test()

	local yourCFrame = CFrame.new(0,5,0)

	local oldtime = os.clock()
	local MineCast = SinlerClyinder:Solve(yourCFrame)

	if MineCast then
		table.insert(SinlerCast, os.clock() - oldtime)
		
	end

	local oldtimed = os.clock()
	local Cast = OrginialClyinder:Solve(yourCFrame)

	if Cast then
		table.insert(OrginialCast, os.clock() - oldtimed)
	end

end


for i = 1, 1000  do
	
	test()
end

table.sort(SinlerCast, function(a,b )
	return a > b
end)
	
table.sort(OrginialCast, function(a,b )
	return a > b
end)

print("SinlerCast:", SinlerCast[math.floor(#SinlerCast * 0.5 )])
print("ClyindricalCast:", OrginialCast[math.floor(#OrginialCast * 0.5 )])
```

**Comment by IdleBrickRBLX** [comment]
This is a great improvement on performance. I was thinking the same thing when you removed the second loop, If you come up with a new method for that or return that part to the old one I will merge the pull request!

File: src/CylindricalCast.lua [hunk 1]
```diff
@@ -17,62 +24,62 @@ type SolverType = {
...
+
+		local position = (newCFrame * CFrame.new(Vector3.new(math.cos(angle) * 0.1, 0, math.sin(angle) * 0.1))).Position
+		local direction = (CFrame.new(position, newCFrame.Position) * CFRAME_ANGLES_AXIS_PI).LookVector
+
+		local raycastResult = Workspace:Raycast(position, direction * ((self.Size.Y * 0.832) + 0.5), self.RaycastParams)
+		
+		
+		if raycastResult then
+			if raycastResult.Instance then
+				table.insert(raycasts, raycastResult)  -- not too sure why this is here, you can just get the first ray that intersected with an object
```
Comment (IdleBrickRBLX) [inline src/CylindricalCast.lua:67]: It works by gathering all the raycasts and gets the raycast thats the closest to the centre of the cylinder

Comment (sinlerdev) [reply src/CylindricalCast.lua:67]: May I ask why you need this? Can't you just return the first ray that was intersected with the object? This way, the ray that you return definitely holds the closest instance to the cylinder.

If I am misunderstanding something, and you seriously need the raycasts table and such, can't we just pre-allocate it on a cylinder constructions? This way, we can avoid memory allocations  every `Solver:solve()` call, which typically get called every frame or so.

**Comment by sinlerdev** [comment]
I have been working on some optimization for my fork *(which will hopefully be shipped to the original one)*, and I have made some trade offs that make sense if you look at it the way I do.

The trade-off I made is to increase the speed of `Solver:Solve` at the cost of `Solver.new`'s speed because if you think about it, cylinders created from the new constructors are just shapes, not objects that were supposed to be related to an individual object *(i.e, creating a cylinder as a hitbox for every player with the same config)*. How I done it is to basically pre-compute all the cframes and positions partially, which comes with an acceptable increase in speed. *(you will understand what I mean when you look at the source code)*

Another thing I made is to reuse the raycasts table between `...:Solve()` calls for the sake of avoiding allocations, moreover, I tried to avoid computing something twice in your code, as thats why there is a variable in the third/second line of the solve function computing `((self.Size.Y * 0.832) + 0.5)`  because I don't think recomputing it for `n` times while `n` is `thickness * quality` will be a great idea.

Finally, I made the `...:Solve()` method return the raycasts table *(the same table that the cylinder object is using, which may cause bugs in the future if the end-user tried to mutate it, but I think freezing the table would be a great fix to this)* because I don't like the fact that I don't have the information about how many objects my rays intersected with, so it is more of a precision increase while having no negative impact over performance. The sorting part is still there but commented out though.
_________________________________

With all the things I have listed, I have seen an acceptable performance increase during testing in the `Solver:Solve()` method, which was what I was planning for. 

Use the rbxl file in this [release](https://github.com/sinlerdev/CylindricalCast/releases/tag/test) for the benchmarking suite I used.

**Comment by IdleBrickRBLX** [comment]
I will test your release and run some benchmarks.

**Comment by IdleBrickRBLX** [comment]
After doing some benchmarks and checking it all works i have concluded that this is ready to be merged with the main branch

**Comment by IdleBrickRBLX** [comment]
Also a tip is `for t,_ in pairs(table.create(self.ThicknessQuality, "")) do` is faster than `t = 1, self.ThicknessQuality` after doing benchmarks for last release

**Comment by IdleBrickRBLX** [comment]
Which can be seen in these benchmarks 
image

**Comment by sinlerdev** [comment]
> Which can be seen in these benchmarks image

That's weird, as in my testing the latter is the faster *(although the difference isn't noticeable that much)*. 
image

code:
```lua
local ReplicatedStorage = game:GetService('ReplicatedStorage')
local RunService = game:GetService("RunService")
local Bencher = require(ReplicatedStorage.Bencher)

local timings_for_numloop = Bencher.measure(500, function()
	for i = 1, 30 do
		print(i * 1)
	end
end)

local timings_for_pair = Bencher.measure(500, function()
	for t,_ in pairs(table.create(30, "")) do
		print(t * 1)
	end
end)

print(Bencher.returnPercentile(timings_for_numloop, 0.5))
print(Bencher.returnPercentile(timings_for_pair, 0.5))

```

**Comment by IdleBrickRBLX** [comment]
yea ipairs is faster than pairs, Make sure to also run your benchmarks a couple times and average the results

**Comment by IdleBrickRBLX** [comment]
Try buying the benchmarker plugin by boatbomber

**Comment by sinlerdev** [comment]
> yea ipairs is faster than pairs, Make sure to also run your benchmarks a couple times and average the results

True, but the `for idex = start, end, increase` loop is the fastest among all of them. Anyway, I think creating a whole table just to not use it is quite unnecessary.

> Try buying the benchmarker plugin by boatbomber

I currently don't have the money for it, but I use a similar method in calculating the 50th percentile.

**Comment by IdleBrickRBLX** [comment]
Ok so i merged the pull request with main and also added you to become a contributor for the repo

**Pull request merged**